node/overseer/src/lib.rs

```diff
@@ -1330,50 +1329,62 @@ where
 		self.broadcast_signal(OverseerSignal::ActiveLeaves(update)).await?;
 
 		loop {
-			while let Poll::Ready(Some(msg)) = poll!(&mut self.events_rx.next()) {
-				match msg {
-					Event::MsgToSubsystem(msg) => {
-						self.route_message(msg).await;
-					}
-					Event::Stop => {
-						self.stop().await;
-						return Ok(());
-					}
-					Event::BlockImported(block) => {
-						self.block_imported(block).await?;
-					}
-					Event::BlockFinalized(block) => {
-						self.block_finalized(block).await?;
-					}
-					Event::ExternalRequest(request) => {
-						self.handle_external_request(request);
-					}
-				}
-			}
-
-			while let Poll::Ready(Some((StreamYield::Item(msg), _))) = poll!(
-				&mut self.running_subsystems_rx.next()
-			) {
-				match msg {
-					ToOverseer::SubsystemMessage(msg) => self.route_message(msg).await,
-					ToOverseer::SpawnJob { name, s } => {
-						self.spawn_job(name, s);
+			select! {
+				msg = self.events_rx.next().fuse() => {
+					let msg = if let Some(msg) = msg {
+						msg
+					} else {
```

Comment:
In fact we probably need the priority of the external events to signal subsystems about deactivated blocks and such

Reply:
Yeah my reasoning was probably bad, however I don't assume that these events will not hit the overseer on time.

Reply:
I don't think this should be a problem in practice. However, a bigger problem is, that we have bounded channels between overseer and subsystems and if a subsystem is busy, this can block overseer from any other action until the subsystem receives a message/signal. Maybe we should consider dropping messages instead.

Reply:
I don't think dropping messages is a good idea. It would be nice to have backpressure on block import instead, but Substrate isn't tooled for that yet.

Reply:
Still, these channels are pretty performant and the amount of messages coming through are relatively low (~hundreds  or low thousands per block). I would be highly surprised if the overseer could not deliver at least an order of magnitude beyond that.